stats/opentelemetry/e2e_test.go

```diff
@@ -582,3 +601,365 @@ func pollForWantMetrics(ctx context.Context, t *testing.T, reader *metric.Manual
+							Key: "message-size-compressed",
+						},
+					},
+				},
+			},
+		},
+		{
+			name:     "Attempt.grpc.testing.TestService.UnaryCall",
+			spanKind: trace2.SpanKindInternal.String(),
+			attributes: []attribute.KeyValue{
+				{
+					Key: "Client",
+				},
+				{
+					Key: "FailFast",
+				},
+				{
+					Key: "previous-rpc-attempts",
+				},
+				{
+					Key: "transparent-retry",
+				},
+			},
+			events: []trace.Event{
+				{
+					Name: "Outbound compressed message",
+					Attributes: []attribute.KeyValue{
+						{
+							Key: "sequence-number",
+						},
+						{
+							Key: "message-size",
+						},
+						{
+							Key: "message-size-compressed",
+						},
+					},
+				},
+				{
+					Name: "Inbound compressed message",
+					Attributes: []attribute.KeyValue{
+						{
+							Key: "sequence-number",
+						},
+						{
+							Key: "message-size",
+						},
+						{
+							Key: "message-size-compressed",
+						},
+					},
+				},
+			},
+		},
+		{
+			name:       "grpc.testing.TestService.UnaryCall",
+			spanKind:   trace2.SpanKindClient.String(),
+			attributes: []attribute.KeyValue{},
+			events:     []trace.Event{},
+		},
+	}
+
+	for index, span := range spans {
+		// Check that the attempt span has the correct status
+		if got, want := spans[index].Status.Code, otelcodes.Ok; got != want {
+			t.Errorf("Got status code %v, want %v", got, want)
+		}
+		// name
+		if got, want := span.Name, wantSI[index].name; got != want {
+			t.Errorf("Span name is %q, want %q", got, want)
+		}
+		// spanKind
+		if got, want := span.SpanKind.String(), wantSI[index].spanKind; got != want {
+			t.Errorf("Got span kind %q, want %q", got, want)
+		}
+		// attributes
+		if got, want := len(span.Attributes), len(wantSI[index].attributes); got != want {
+			t.Errorf("Got attributes list of size %q, want %q", got, want)
+		}
+		for idx, att := range span.Attributes {
+			if got, want := att.Key, wantSI[index].attributes[idx].Key; got != want {
+				t.Errorf("Got attribute key for span name %q as %q, want %q", span.Name, got, want)
+			}
+		}
+		// events
+		if got, want := len(span.Events), len(wantSI[index].events); got != want {
+			t.Errorf("Event length is %q, want %q", got, want)
+		}
+		for eventIdx, event := range span.Events {
+			if got, want := event.Name, wantSI[index].events[eventIdx].Name; got != want {
+				t.Errorf("Got event name for span name %q as %q, want %q", span.Name, got, want)
+			}
+			for idx, att := range span.Attributes {
+				if got, want := att.Key, wantSI[eventIdx].attributes[idx].Key; got != want {
+					t.Errorf("Got attribute key for span name %q with event name %q, as %q, want %q", span.Name, event.Name, got, want)
+				}
+			}
+		}
+	}
+}
+
+// TestW3CContextPropagator sets up a stub server with  OpenTelemetry tracing
+// enabled makes a unary and a streaming RPC, and then, asserts that the correct
+// number of spans are created with the expected spans.
+func (s) TestW3CContextPropagator(t *testing.T) {
```

Comment:
This can be just `TestSpan_WithW3CContextPropagator` and do all the things as `TestSpan` or can be combined with TestSpan with only difference being setting the tracepropagator and everything should remains same.